File: binding/src/libqgenericplugin.cpp
#include <QGenericPlugin>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qgenericplugin.h>
#include "libqgenericplugin.h"
#include "libqgenericplugin.hxx"

QGenericPlugin* QGenericPlugin_new() {
	 return new VirtualQGenericPlugin();
}

QGenericPlugin* QGenericPlugin_new2(QObject* parent) {
	 return new VirtualQGenericPlugin(parent);
}

libqt_string QGenericPlugin_Tr(const char* s) {
	QString _ret = QGenericPlugin::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QObject* QGenericPlugin_Create(QGenericPlugin* self, const libqt_string name, const libqt_string spec) {
	return self->create(QString::fromUtf8(name.data, name.len), QString::fromUtf8(spec.data, spec.len));
}

libqt_string QGenericPlugin_Tr2(const char* s, const char* c) {
	QString _ret = QGenericPlugin::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QGenericPlugin_Tr3(const char* s, const char* c, int n) {
	QString _ret = QGenericPlugin::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
QObject* QGenericPlugin_QBaseCreate(QGenericPlugin* self, const libqt_string name, const libqt_string spec) {
	auto* vqgenericplugin = dynamic_cast<VirtualQGenericPlugin*>(self);
	if (vqgenericplugin && vqgenericplugin->isVirtualQGenericPlugin) {
vqgenericplugin->setQGenericPlugin_Create_IsBase(true);
	return vqgenericplugin->create(QString::fromUtf8(name.data, name.len), QString::fromUtf8(spec.data, spec.len));
}
}

// Auxiliary method to allow providing re-implementation
void QGenericPlugin_OnCreate(QGenericPlugin* self, intptr_t slot) {
	auto* vqgenericplugin = dynamic_cast<VirtualQGenericPlugin*>(self);
	if (vqgenericplugin && vqgenericplugin->isVirtualQGenericPlugin) {
vqgenericplugin->setQGenericPlugin_Create_Callback(reinterpret_cast<VirtualQGenericPlugin::QGenericPlugin_Create_Callback>(slot));
}
}

void QGenericPlugin_Delete(QGenericPlugin* self) {
    delete self;
}

File: binding/src/libqgenericplugin.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qgenericplugin.html
pub const qgenericplugin = struct {

    /// New constructs a new QGenericPlugin object.
    pub fn New() QtC.QGenericPlugin {
        return qtc.QGenericPlugin_new();
    }


    /// New2 constructs a new QGenericPlugin object.
    pub fn New2(parent: ?*anyopaque) QtC.QGenericPlugin {
        return qtc.QGenericPlugin_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGenericPlugin_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qgenericplugin.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Create(self: ?*anyopaque, name: []const u8, spec: []const u8) QtC.QObject {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const spec_str = qtc.libqt_string{
    .len = spec.len,
    .data = spec.ptr,
};
return qtc.QGenericPlugin_Create(@ptrCast(self), name_str, spec_str);
    }

    /// Allows for overriding the related default method
    pub fn OnCreate(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, [*:0]const u8) callconv(.c) QtC.QObject) void {
        qtc.QGenericPlugin_OnCreate(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseCreate(self: ?*anyopaque, name: []const u8, spec: []const u8) QtC.QObject {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
const spec_str = qtc.libqt_string{
    .len = spec.len,
    .data = spec.ptr,
};
return qtc.QGenericPlugin_QBaseCreate(@ptrCast(self), name_str, spec_str);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGenericPlugin_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qgenericplugin.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGenericPlugin_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qgenericplugin.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QGenericPlugin_Delete(@ptrCast(self));
    }
};
